Homework lesson 3

woodward42 opened the pull request.

woodward42 committed:
Переписал на классы по аналогии с уроком

woodward42 committed:
ДЗ#2. Добавил классы с описанием + метод суммы

woodward42 committed:
Задачка по гамбургеру готова. В папке проекта

woodward42 committed:
Переписал main, дз xhr => Promise готово

woodward42 committed:
Стили корзины, добавлен рендер по типу товара

woodward42 committed:
Добавлен класс родитель, классы каталог и корзина

Товары каталога и корзины получаем из JSON, все ок

Доделать методы и передачу корзины в ProductsList

woodward42 committed:
Косметика

woodward42 committed:
Готов метод добавления товара в корзину

woodward42 committed:
Убрал лишнее

woodward42 committed:
Готовы методы добавления и удаления для корзины

AndrewRimsky commented on project/js/main.js:
```diff
@@ -1,32 +1,177 @@
+const API = 'https://raw.githubusercontent.com/woodward42/online-store-api/master/responses';
+
+//обработчик клика по иконке корзины, чтобы показать/скрыть список товаров корзины
+const cartProducts = document.querySelector('.cart-products')
+      document.querySelector('.cart-btn').addEventListener('click', () => {
+          cartProducts.classList.toggle('shown')
+      })
```
Это лучше утащить в корзину, зачем оно нам здесь?

AndrewRimsky commented on project/js/main.js:
```diff
@@ -1,32 +1,177 @@
+      })
+
+
+//универсальный класс товара, рендерится разная вёрстка, в зависимости от переданного из родителя (ProductsList или CartList) типа
```
Вот это странноватая история. Неопределенный объект который то такой, то иной это не очень здорово. Лучше было просто сделать цепочку наследования (Базовый товар с конструктором от него наследуются классы Товар корзины и Товар каталога)

AndrewRimsky commented on project/js/main.js:
```diff
@@ -1,32 +1,177 @@
+                        <h3 class="product-title">${item.product_title}</h3>
+                        <img class="product-img" src="${item.product_image}" alt="img" width="200">
+                        <p class="product-price">${item.product_price}$</p>
+                        <button 
+                            class="product-add-to-cart-btn" 
+                            name="add"
+                            data-id="${item.product_id}" 
+                            data-title="${item.product_title}" 
+                            data-image="${item.product_image}" 
+                            data-price="${item.product_price}"
+                            >Add to Cart</button>
```
Да, в принципе такой вариант тоже имеет место быть

AndrewRimsky commented on project/js/main.js:
```diff
@@ -1,32 +1,177 @@
-    //метод map возвращает новый массив, каждый элемент productsList передаётся в ф-ю renderProduct и там распарсится в вёрстку
-    const productsElements = productsList.map(item => renderProduct(item)).join(' ');
-            document.querySelector('.products').innerHTML = productsElements;
+//создаем общий класс, от которого будут наследоваться классы список товаров и список корзины
```
Тут аналогично как и с классом элемента надо поступить. Общий родитель от него наследуем более конкретные классы

AndrewRimsky commented on project/js/main.js:
```diff
@@ -1,32 +1,177 @@
+        this.container = document.querySelector(container)
+        this.type = type
+        this.url = `${API}${url}`
+        this.cart = cart
```
Корзина есть у всех, даже у корзины получается) Вот это и не здорово, в объекте не должно быть того, что ему не надо

AndrewRimsky commented on project/js/main.js:
```diff
@@ -1,32 +1,177 @@
+    _renderProductsList(){  //я переписал этот метод, потому что у меня не получилось без перерендера при добавлении товара корректно отрисовать корзину
+        let htmlString = ''
+        for (let product of this.products){
+            
+            if (!this.isRendered){
+                htmlString += product.renderProduct() //вставили в DOM дерево в контейнер
+            }
+            else continue;
```
Это лишнее. И так ничего не происходит после блока иф, поэтому элс можно удалить. Если хотите явно отсекать по условию и не делать ничего для отрендеренных товаров, то можно писать так
```
if (this.isRendered) {
  continue;
}

htmlString += product.renderProduct()
```

AndrewRimsky commented on project/js/main.js:
```diff
@@ -1,32 +1,177 @@
+
+    //функция аналогичная функции в корзине, вызывать её попробую после рендера в родительском классе List
+    _addCatalogListeners(){
+        this.container.addEventListener('click', evt => {   //обработчик вешаем на контейнер, клики поймаем через вспылтие
```
Отлично! Это очень хороший вариант, чтобы не вешать обработчик на каждую конкретную кнопку

AndrewRimsky commented on project/js/main.js:
```diff
@@ -1,32 +1,177 @@
+    }
+    _remove(id){
+        //находим в каждом из массивов по id элемент, который надо удалить
+        let productToDeleteFromData = this.data.find(item => item.product_id == id)
```
Зачем из данных удалять?

AndrewRimsky commented on project/xhr_to_promise.html:
```diff
@@ -0,0 +1,56 @@
+const xhrUrl = 'https://raw.githubusercontent.com/woodward42/online-store-api/master/responses/catalogData.json'
+getDataXHR( xhrUrl, data => console.log(`From getDataXHR:`,JSON.parse(data)) )
+
+//теперь то же самое, но в Promise
```
Да, совершенно верно

AndrewRimsky commented on project/js/main.js:
```diff
@@ -1,32 +1,177 @@
+            }
+            else continue;
+        }
+        this.container.innerHTML = htmlString
```
Да, вот постоянный перерендер наверное не самый оптимальный вариант, но в целом для нас сейчас не очень критично это

AndrewRimsky commented on project/js/main.js:
```diff
@@ -1,32 +1,177 @@
+        })
+    }
+
+    add(item){
```
лучше addItem (и remove тоже в removeItem) переименовать. И сам метод сейчас постоянно создает новые позиции в корзине, тут стоит пересмотреть подход - смотреть есть ли уже товар в корзине, если есть, то новый экземляр же нам и не нужен. Надо менять кол-во уже имеющегося в корзине товара. Также и с удалением, надо менять кол-во если уже в корзине 2 и более единиц одного товара, или удалять из массива, если товар один

AndrewRimsky reviewed (approved):
Все отлично!

woodward42 committed:
//переделать API
//добавить метод json при добавлении и удалении
//сделать удаление при 1 товаре в корзине

woodward42 committed:
Доделал.
Убрал перерендер при удалении, добавил обновление элемента в корзине
Добавил запросы к серверу при добавлении и удалении
